prow/spyglass/gcsartifact.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"compress/gzip"
+	"context"
+	"io/ioutil"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+// GCSArtifact represents some output of a prow job stored in GCS
+type GCSArtifact struct {
+	// The handle of the object in GCS
+	handle *storage.ObjectHandle
+
+	// The direct link to the Artifact, can be used for read operations
+	link string
+
+	// The path of the Artifact within the job
+	path string
+
+	// Max size to read before failing
+	sizeLimit int64
+}
+
+// NewGCSArtifact returns a new GCSArtifact with a given handle, canonical link, and path within the job
+func NewGCSArtifact(handle *storage.ObjectHandle, link string, path string, sizeLimit int64) *GCSArtifact {
+	return &GCSArtifact{
+		handle:    handle,
+		link:      link,
+		path:      path,
+		sizeLimit: sizeLimit,
+	}
+}
+
+// Size returns the size of the artifact in GCS
+func (a *GCSArtifact) Size() int64 {
+	attrs, err := a.handle.Attrs(context.Background())
+	if err != nil {
+		logrus.WithError(err).Error("Could not retrieve object attributes")
+	}
+	return attrs.Size
+}
+
+// JobPath gets the GCS path of the artifact within the current job
+func (a *GCSArtifact) JobPath() string {
+	return a.path
+}
+
+// CanonicalLink gets the GCS web address of the artifact
+func (a *GCSArtifact) CanonicalLink() string {
```

Comment:
Why does everything need accessor methods?

Reply:
In the artifact case, to ensure that viewers have access to some consistent data about any given artifact.

Reply:
Consider starting everything unexported an then exporting them only when necessary in subsequent PRs.

Reply:
Fair enough. In general I would recommend waiting until the PR that uses them to export (this simplifies review).

Otherwise I cannot distinguish between "this needs to be exported" from "I'm exporting by default because I'm used to another language." This seems important given that this seemed to start with everything exported